### ISSUE TITLE
Fetchers failing for fastjet

### ISSUE THREAD

**Issue by vvolkl** [LLNL/spack#20050]
I tried to add a new version for  [fastjet](https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/fastjet/package.py) but the fetcher has an issue (the fastjet release / download page is still working and unchanged, as far as I can tell: http://fastjet.fr/all-releases.html)

### Steps to reproduce the issue

```console
spack checksum fastjet

==> Something failed on http://fastjet.fr/repo/fastjet-3.3.3.tar.gz, skipping.  (All fetchers failed)

```

**Comment by vvolkl** [comment]
The installation is working, as long as the versions are taken from the source mirror (which is why this might have been failing for a while but wasn't noticed)

**Comment by alalazo** [comment]
> The installation is working, as long as the versions are taken from the source mirror

Are you referring to a local mirror you have on your machine vs. `spack-public` or to a the package `url` being correct and `spack-public` being not?

EDIT: Replying to my own question:
```
$ spack fetch fastjet
==> Fetching https://spack-llnl-mirror.s3-us-west-2.amazonaws.com/_source-cache/archive/30/30b0a0282ce5aeac9e45862314f5966f0be941ce118a83ee4805d39b827d732b.tar.gz
########################################################################################################################################### 100,0%
[ Remove cache, use default url ]

$ spack fetch fastjet
==> Error: All fetchers failed
```

**Comment by vvolkl** [comment]
yes, I meant that as long as fastjet is fetched from any existing source mirror, it can be installed, so the fetcher must have broken after the fastjet sources were added to the public source mirror.

**Comment by alalazo** [comment]
The issue apparently comes from this function:

https://github.com/spack/spack/blob/b490d65f28e2f746bb93746ff854afd299881cdd/lib/spack/spack/fetch_strategy.py#L321-L328

which is used to understand if the URL exists. Fastjet fails on it for some reason (exit code 22, which means HTTP error 400 or above) so fetching the whole tarball is not even tried. Need to look at this more closely to decide what is the best solution, but it seems the issue is affecting only `fastject` (or a small number or packages).

**Comment by vvolkl** [comment]
Indeed it is a 403:
```
<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html><head>
<title>403 Forbidden</title>
</head><body>
<h1>Forbidden</h1>
<p>You don't have permission to access /repo/fastjet-3.3.4.tar.gz
on this server.</p>
<hr>
<address>Apache Server at fastjet.fr Port 80</address>
</body></html>
```

For me it works when adding `'-O'` to `curl_args`, but I don't really understand why

**Comment by vvolkl** [comment]
Strange, looking at this site: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Ranges

it seems like the server *should* accept range requests.

```
~$ curl -I   http://fastjet.fr/repo/fastjet-3.3.4.tar.gz
HTTP/1.1 200 OK
Date: Mon, 23 Nov 2020 22:10:44 GMT
Server: Apache
Last-Modified: Tue, 05 May 2020 17:13:31 GMT
ETag: "3400003-2c6c47-5a4e9c4539290"
Accept-Ranges: bytes
Content-Length: 2911303
Connection: close
Content-Type: application/x-gzip
```

Using a header request (`curl -I`) might actually accomplish the same thing (will return a 404 if the url is not valid)

EDIT: @alalazo  maybe `_existing_url` can try the header request if the range request fails, and only return failure if both of them fail?

**Comment by vvolkl** [comment]
@alalazo #20064 seems like a fairly straightforward fix that just makes fetching a bit more robust, what do you think?
